src/Compilers/Server/VBCSCompiler/NamedPipeClientConnectionHost.cs

```diff
@@ -94,7 +94,26 @@ public void EndListening()
             Debug.Assert(_queue is object);
             Debug.Assert(_listenTasks is object);
 
-            _cancellationTokenSource.Cancel();
+            try
+            {
+                // Even though the Tasks created to run the compilation servers can never throw, 
+                // the CancellationToken from this source ends up getting passeed throughout the 
+                // named pipe infrastructure. Parts of that infrastructure hook into 
+                // CancellationToken.Register and those will throw during a Cancel operation. 
+                //
+                // Most notably of these is IOCancellationHelper.Cancel. This has a race where it
+                // will try to cancel IO on a disposed SafeHandle. That causes an ObjectDisposedException
+                // to propagate out from the Cancel method here.
+                //
+                // There is no good way to guard against this hence we just have to accept it as a
+                // possible outcome.
+                _cancellationTokenSource.Cancel();
```

Comment:
❔ Why not use `CancelAfter(0)` to run the cancellation continuations asynchronously?

Reply:
I would rather distinguish between the two scenarios:

1. `Cancel` throwing is generally indicative of a runtime bug. 
1. `WaitAll` throwing should never happen. That represents a bug in our code that needs immediate attention.